lib/asciidoctor/document.rb

```diff
@@ -383,16 +383,20 @@ def initialize data = nil, options = {}
     # legacy support for numbered attribute
     attr_overrides['sectnums'] = attr_overrides.delete 'numbered' if attr_overrides.key? 'numbered'
 
-    # If the base_dir option is specified, it overrides docdir and used as the root for relative
-    # paths. Otherwise, the base_dir is the directory of the source file (docdir), if set, or else
+    # If the base_dir option is specified, it overrides docdir and is used as the root for relative
+    # paths. Otherwise, the base_dir is the directory of the source file (docdir), if set, otherwise
     # the current directory.
-    if options[:base_dir]
-      @base_dir = attr_overrides['docdir'] = ::File.expand_path(options[:base_dir])
+    if (base_dir_val = options[:base_dir])
+      if ::RUBY_ENGINE_OPAL && ::JAVASCRIPT_IO_MODULE == 'xmlhttprequest' && (Helpers.uriish? base_dir_val)
```

Comment:
I think this change is too conservative.
`File.expand_path` does not play well with URIs and we will have issues down the line:

```ruby
p Dir.pwd # /home/guillaume
p File.expand_path('file:///path/to/asciidoctor.js') # /home/guillaume/file:/path/to/asciidoctor.js
p File.expand_path('https://asciidoctor.org') # /home/guillaume/https:/asciidoctor.org
```

In my opinion, if `base_dir_val` is an URI we should not use `File.expand_path` no matter what the runtime is.

Reply:
I agree with you, but we're not ready for that until #2592, which is a much larger change. Right now we are focused just on the xmlhttprequest environment.

Outside of the xmlhttprequest environment, base_dir must be an absolute path. This is a deep assumption in Asciidoctor that is going to take some effort to weed out.

Reply:
(we can assume it's okay in the browser environment since it doesn't do system_path operations).

Reply:
> I agree with you, but we're not ready for that until #2592, which is a much larger change. Right now we are focused just on the xmlhttprequest environment.

Makes sense.

> (we can assume it's okay in the browser environment since it doesn't do system_path operations).

I think it's not needed because this is an intermediate step and browser environment almost always use xmlhttprequest anayway

Reply:
When we do fix #2592, then we'll store a URI value as a URI type. That way, we can easily detect what type of value we have an act appropriately. That will be something we'll do in 1.5.7.

Reply:
What we're saying is that until #2592, only the browser environment (xmlhttprequest) is allowed to set the base_dir to a URI (URL or file://). We may or may not need it, but I think it's worth starting there at least.

Reply:
Note that with the current changes in both branches, the other option is to set the docdir attribute from the API. That value is not modified. That may even void the need for the exception in the base_dir logic. It was needed for Antora, which has a virtual filesystem and therefore expanding docdir against the physical filesystem makes no sense.